### PR TITLE
Reduce the work to add a new expr: unify the structure of exprs

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I added a new data member to `Expr`: `std::vector<Val*> attributes_`. Subclasses of `Expr` are now not allowed to have their own data member in the class. Instead, if a data member is neither an input nor an output, then the subclass should register it as an attribute. Some subclass of  `Expr` needs to store plain data that is not a `Val`, such as `DataType`, `bool`, etc. These plain data are not `Val`, therefore impossible to store as an attribute. To work around this, I created a class `template <typename T> class Attribute : public Val`  which can be used to store data of type `T`.

By following the contract that subclass of `Expr` must store data inside inputs, outputs, or attributes, every expr type can be constructed with the following ctor:
```C++
Expr(IrBuilderPasskey, std::vector<Val*> inputs, std::vector<Val*> outputs, std::vector<Val*> attributes);
```
And IR cloning and shallow copying now become independent of expr type. Now adding a new expr type does not need to manually write down the cloning constructor, `sameAs`, and `shallowCopy`. They can all be inherited from `Expr`. This is a big save that helps me to reduce ~800 lines of code.

I also refactored `IrCloner` to move the type dispatch as a member function.

I believe that with my change in this PR, we can also greatly simplify IR mutator. But I will not do it in this PR.

I recommend starting the review from `torch/csrc/jit/codegen/cuda/ir_base_nodes.h`.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_builder.cpp [hunk 1]
```diff
@@ -8,74 +8,6 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
-//! Clone an IR node, forwarding the arguments to the IrCloner constructor.
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_builder.cpp:14]: moved to header file

File: torch/csrc/jit/codegen/cuda/ir_builder.h [hunk 2]
```diff
@@ -15,20 +16,6 @@ class Kernel;
 
 class IrCloner;
 
-// Passkey for builder to register properties with statements, and to call
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_builder.h]: Moved to a separate file `torch/csrc/jit/codegen/cuda/ir_builder_key.h` so it can be included in `torch/csrc/jit/codegen/cuda/ir_base_nodes.h`.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 3]
```diff
@@ -436,12 +474,25 @@ class TORCH_CUDA_CU_API Expr : public Statement {
...
+  Expr(
+      IrBuilderPasskey,
+      std::vector<Val*> inputs,
+      std::vector<Val*> outputs,
+      std::vector<Val*> attributes);
+
   // Creates a new instance of the expression with all its field copied.
   // Note that unlike IrCloner, this function only do a shallow copy
-  virtual Expr* shallowCopy() const = 0;
+  Expr* shallowCopy() const;
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h:485]: This method now is not even virtual

File: torch/csrc/jit/codegen/cuda/ir_builder.h (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_builder.h]: Oh, I did that too in #2197 😄

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 4]
```diff
@@ -70,6 +71,14 @@ class ExprPasskey {
 
 TORCH_CUDA_CU_API void swap(Fusion& a, Fusion& b) noexcept;
 
+#define DECLARE_CLONE \
+  virtual Statement* clone(IrCloner* ir_cloner) const override;
+
+#define DEFINE_CLONE(ClassName)                            \
+  Statement* ClassName::clone(IrCloner* ir_cloner) const { \
+    return IrBuilder::clone(this, ir_cloner);              \
+  }
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Can we undef these somewhere? Otherwise, we should use names like `PYTORCH_NVFUSER_DECLARE_CLONE`.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h [hunk 5]
```diff
@@ -391,6 +404,31 @@ class TORCH_CUDA_CU_API Val : public Statement {
   int evaluator_index_ = -1;
 };
 
+//! A Val object that stores a plain data. Note that this class is only intended
+//! to hold non-IR data, such as DataType, std::vector<int>, etc. Please don't
+//! use this class to hold IR nodes or their pointers.
+template <typename T>
+class TORCH_CUDA_CU_API PlainVal : public Val {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: So, this is intended to hold non-IR data, but then why is it a subclass of `Val`?

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 6]
```diff
@@ -32,64 +32,55 @@ bool areEqualScalars(Val* v1, Val* v2);
 
 class TORCH_CUDA_CU_API FullOp : public Expr {
  public:
-  FullOp(IrBuilderPasskey, Val* out, Val* fill_value, DataType dtype);
+  using Expr::Expr;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:35]: Why is this needed?

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: So, this is used to hold IR node attributes. Does it need to be a subclass of `Val`?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Because `attributes_` are vectors of `Val*` and I want to push plain values into `attributes_`. Maybe I should change `attributes_` to a vector of `Statement*`? Currently, lots of kernel IR has expr members, and I am storing them as something like `PlainVal<Statement*>`; maybe making `attributes_` a vector of `Statement*` would make it easier because I can directly store it.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h [hunk 7]
```diff
@@ -513,61 +465,51 @@ class TORCH_CUDA_CU_API ReductionOp : public Expr {
...
   //! iteration grouping.
   size_t numExprs() const {
-    return reduction_op_types_.size();
+    return getReductionOpTypes().size();
   }
 
-  const std::vector<Val*>& initVals() const {
-    return init_vals_;
+  std::vector<Val*> initVals() const {
+    return {attributes().begin() + 2, attributes().end()};
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_internal_nodes.h]: A bit concerning that we need to make sure we use correct offsets in the attribute vector. It seems like we are effectively losing static type safety, although the overall IR classes are more consistent.

Given that typically there are not many attributes in each class, I don't think the above concern is critical.

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: I don't think so. These macros will be used in cpp files. I will rename it as `NVFUSER_DECLARE_CLONE`.

File: torch/csrc/jit/codegen/cuda/ir_internal_nodes.h (hunk 6 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:35]: It is for the following ctors:
```
Expr(IrBuilderPasskey, std::vector<Val*> inputs, std::vector<Val*> outputs, std::vector<Val*> attributes);
Expr(const Expr* src, IrCloner* ir_cloner);
```

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Ah, I see. Yeah, we want to store proper IR-node values as well. Can it be `std:vector<std::any>`?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Not today. `std::any` is C++17 but PyTorch now is still using C++14

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: OK, but my question still holds. Can we just have something like:

```
class AbstractExprAttribute {
};

template <typename T>
class ExprAttribute: AbstractExprAttribute {
   T data_;
   ...
};
```

Then in `Expr`, we could have `std::vector<AbstractExprAttribute> attributes_`.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: I think it can only be `std::vector<std::shared_ptr<AbstractExprAttribute>> attributes_`?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Ah, right. Can it be a vector of `unique_ptr`? 

I just feel odd to have arbitrary attributes, like a `DataType` and a bool flag, as a `Val`. Please convince me if you really want to use `PlainVal`.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: I believe the biggest advantage of making it a `Val` is that, `IrCloner` will automatically help you do deep copy when necessary. For example, if I have an attribute `bool`, and I clone the expr, then I will also have a new `PlainVal<bool>` object for the new expr. If later, I changed the new expr's bool attribute, the old one's bool attribute is not changed. But if we manage this by hand, we could make mistakes that, when cloning an expr, the old and new expr are pointing to the same object for its attribute, and if we change one, the other is also changed.

So to answer your question:
> Can it be a vector of `unique_ptr`?

Yes. Actually, it SHOULD be a `unique_ptr` instead of a `shared_ptr`, for the reason I described above.

But without using `PlainVal`, we will have to manually do a deep copy of the `std::vector<std::unique_ptr<AbstractExprAttribute>> attributes_`. It's a little inconvenient, but should be more efficient.

I don't have a strong preference about which one we should choose.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Oh, wait, no. It is not "a little inconvenient", but super convenient. Without knowing `T`, we can not do a deep copy.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: OK, that makes sense. (I assume you meant "super inconvenient")

Let's use `PlainVal`, but I'd prefer a name like `Attribute`, which I think a little more clearly implies what it's supposed to be used for.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Renamed

File: torch/csrc/jit/codegen/cuda/ir_base_nodes.h (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/ir_base_nodes.h]: Renamed

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for the refactoring.
